File: src/bindings/ruby/Master.rb
require 'msg'
require 'RubyProcess'

include MSG

class Master < RbProcess  
  def initialize2()
    super()
  end
 
  
  #for Testing
  def msg_main2(args)
   info("Hello From Master")
  end
  
  
  # msg_main : that function that will be executed when Running Simulation
  def msg_main(args) # args is an array Containin' arguments for function master
    info("Hello From Master")
    size = args.size
    info ("Number of Args for Master = " + size.to_s)
   for i in 0..size-1
      info(args[i])
   end
   
   raise "Master needs 3 arguments" if size < 3 
   numberOfTask = Integer(args[0]) 
   taskComputeSize = Float(args[1])
   taskCommunicationSize = Float(args[2])
   slaveCount = Integer(args[3]) 
   
   #Creating & Sending Task
   for i in 0..numberOfTask-1
  
     
     task = RbTask.new("Task_"+ i.to_s, taskComputeSize , taskCommunicationSize );
     s_alias = "slave>>" + (i%slaveCount).to_s
     info("Master Sending "+ RbTask.name(task) + " to " + s_alias + " with Comput Size " + RbTask.compSize(task).to_s)
     RbTask.send(task,s_alias)
     info("Master Done Sending " +RbTask.name(task) + " to " + s_alias)
#       sameTask = RbTask.receive(s_alias)
#      puts "Master Receiving its Own Task"
   end
  
   # Sending Finalize Tasks
   info ("Master: All tasks have been dispatched. Let's tell everybody the computation is over.")
   for i in 0..slaveCount-1
     s_alias = "slave " + i.to_s
     info ("Master Sending Finalize to " + s_alias)
     RbTask.send(RbTask.new("finalize",0,0),s_alias)
   end
   info("Master : Everything's Done")
  end  
  
end

File: examples/lua/tracing/master.lua
--Master Function
function Master(...) 

simgrid.info("Hello from lua, I'm the master")
for i,v in ipairs(arg) do
    simgrid.info("Got "..v)
end

prop_value = simgrid.host.get_prop_value(simgrid.host.self(),"peace");
simgrid.info("Prop Value >>> ".. prop_value);

nb_task = arg[1];
comp_size = arg[2];
comm_size = arg[3];
slave_count = arg[4]

if (#arg ~= 4) then
    error("Argc should be 4");
end
simgrid.info("Argc="..(#arg).." (should be 4)")

-- Dispatch the tasks

for i=1,nb_task do
  tk = simgrid.task.new("Task "..i,comp_size,comm_size);
  alias = "slave "..(i%slave_count);
  -- Set Trace Category
  simgrid.Trace.setTaskCategory(tk,"compute");
  simgrid.info("Master sending  '" .. simgrid.task.get_name(tk) .."' To '" .. alias .."'");
  simgrid.task.send(tk,alias); -- C user data set to NULL
  simgrid.info("Master done sending '".. simgrid.task.get_name(tk) .."' To '" .. alias .."'");
end

-- Sending Finalize Message To Others

simgrid.info("Master: All tasks have been dispatched. Let's tell everybody the computation is over.");
for i=0,slave_count-1 do
  alias = "slave "..i;
  simgrid.info("Master: sending finalize to "..alias);
  finalize = simgrid.task.new("finalize",comp_size,comm_size);
  --set Trace Category 
  simgrid.Trace.setTaskCategory(finalize,"finalize");
  simgrid.task.send(finalize,alias);
end
  simgrid.info("Master: Everything's done.");
end

--end of master
